Add optional OpenID Connect authentication

mprahl opened the pull request:
This addresses FACTORY-2993.

yashvardhannanavati commented:
Requesting a review from @accorvin here :)

accorvin commented on src/app/app-routing.module.ts:
```diff
@@ -13,15 +14,22 @@ const routes: Routes = [
+    canActivate: [AuthGuardService]
+  },
+  {
+    path: '**',
```
It might be handy to add a comment describing why this redirect rule is necessary.

accorvin commented on src/app/navbar/navbar.component.ts:
```diff
@@ -1,11 +1,21 @@
-  constructor() { }
+  logout(event: Event): void {
+    this.auth.logOut();
+    event.preventDefault();
```
Maybe a comment about why this preventDefault() call is necessary would be helpful as well

mprahl replied:
Ah, the preventDefault is just to prevent the default behavior of clicking a link. Basically, since the link just executes JS and the JS handles the redirect, we don't want the link to execute it's default behavior of going to the value in the `href` attribute, which is just the convention `#`.

accorvin replied:
Ah ok that makes sense.

mprahl commented:
I'll have Alex from the automation team review this as well since the code is based off his in another project.

amisstea commented on src/app/services/auth-guard.service.ts:
```diff
@@ -0,0 +1,30 @@
+    // a valid access token. This causes the user to be redirected to the OpenID
+    // Connect provider to login. Once the login is complete, the user is redirected
+    // to the requested page that required auth.
+    this.auth.initImplicitFlow(state.url);
```
You might want to consider encoding the url here (perhaps base64) because `angular-oauth-oidc` might attempt to parse the current url by `?` after the provider redirects back to your app. This will be a problem if you ever trigger the implicit flow from a page which has query parameters.

mprahl commented:
I addressed the comment made by @amisstea and I fixed a bug that caused the token to not be received when accessing a protected route directly (not navigating to it from an unprotected route). This was just removing a few lines of code. I don't think this needs another review, so I'll just merge it.
